cmd/thanos/sidecar.go

```diff
@@ -49,6 +50,30 @@ func registerSidecar(m map[string]setupFunc, app *kingpin.Application, name stri
 	gcsBucket := cmd.Flag("gcs.bucket", "Google Cloud Storage bucket name for stored blocks. If empty sidecar won't store any block inside Google Cloud Storage").
 		PlaceHolder("<bucket>").String()
 
+	s3Bucket := cmd.Flag("s3.bucket", "S3-Compatible API bucket name for stored blocks.").
+		PlaceHolder("<bucket>").String()
+
+	s3Endpoint := cmd.Flag("s3.endpoint", "S3-Compatible API endpoint for stored blocks.").
+		PlaceHolder("<api-url>").String()
+
+	s3AccessKey := cmd.Flag("s3.access-key", "Access key for an S3-Compatible API.").
+		PlaceHolder("<key>").String()
+
+	s3SecretKey := cmd.Flag("s3.secret-key", "Secret key for an S3-Compatible API.").
+		PlaceHolder("<key>").String()
+
+	s3Insecure := cmd.Flag("s3.insecure", "Whether to use an insecure connection with an S3-Compatible API.").
+		Default("false").Bool()
+
+	// something abut kingpin (I assume) makes it impossible to dereference and have them resolve correctly later
```

Comment:
This is standard behavior of any flag package. The above just registers flags and the returned pointers get populated once you explicitly parse the flags.
The run function we add to `m` below is then actually called after `main` successfully parsed the flags. So within that function, you can safely derefernce the pointers.

If you move the config struct creation into there, you can use regular strings.

Generally we've tried to avoid creating config structs in Thanos since they make us touch more places on any change without real benefit. But no strict requirement – might be saner here.

Reply:
Fair enough, I can definitely move it down in there. Perhaps we collapse the flags as I mentioned above which would make this easier. Or I can not create the struct, it doesn't make a huge difference to me. 👍